### ISSUE TITLE
Cordova: signInWithRedirect() returns null on iOS but authenticates on Android.

### ISSUE THREAD

**Issue by chrisunderdown** [firebase/firebase-js-sdk#1555]
I followed this guide [https://firebase.google.com/docs/auth/web/cordova](https://firebase.google.com/docs/auth/web/cordova) to use Twitter Authentication in a Cordova app. All was working initially, but now authentication in iOS isn't working.

It successfully redirects back to the app after using the inappbrowser plugin but the `signInWithRedirect()` returns `{"user": null}` most of the time and sometimes does actually get the user object. To clarify, it works on Android every time and returns the user object. 

```javascript
firebase.auth().signInWithRedirect(provider).then(function() {
    return firebase.auth().getRedirectResult()
}).then(function (result) {

    var user = result;
    var token = result.credential.accessToken;
    var secret = result.credential.secret;

    console.log(user);
    // returns {'user': null}

}).catch(function (error) {
    console.log(error)
    console.log('Error Code: ' + error.code);
    console.log('Error Message: ' + error.message);
    // error.code returns undefined
});
```
**Versions:**
iOS: 12.1
Firebase: 5.8.4
Cordova: 8.1.2

**Comment by wti806** [comment]
@chrisunderdown You mentioned it was working initially, did you change anything like upgrading the sdk version or cordova ios version after?

**Comment by premtemp1** [comment]
@chrisunderdown ..

sorry not sure if this correct method to ask...

I have same configuration as you but I get "about:blank"screen after authenticating and it does not go back to my app.. However, if I reopen that app again it works  Can you let me know how you got it work..

Thanks

**Comment by sahyun1** [comment]
having the same issue, works on android but not on iOS

**Comment by chrisunderdown** [comment]
I ended up building a custom auth flow.
Here's a good example of [Instagram Auth](https://github.com/FirebaseExtended/custom-auth-samples/tree/master/instagram) for Firebase.

Also refer to Firebase Docs [signInWithCustomToken](https://firebase.google.com/docs/auth/web/custom-auth)

Hope that helps!

**Comment by ancrezzi** [comment]
Hi!
I'm having a very similar problem: the social login works on Android but not on iOS.

I have an Ionic+Cordova app (Ionic v.3, Cordova v.8.1.2) and I am trying to implement social authentication (Google, Twitter and Facebook) but sometimes I get a blank screen in the redirect.
"Sometimes" because there are times where the functionality works, but these are very random.
This happens only on iOS: i tried on iOS 11.3.1 - 11.4 - 12.3.1; the same code on Android works without problems.

I followed the instructions from here: https://firebase.google.com/docs/auth/web/cordova
I used "signInWithRedirect" because "signInWithPopup" is not supported in Cordova.

Basically, this is what happens:

1) When the user chooses the social login he is redirected to the social page on a different web view;
2) The user enters the login credentials for the selected provider;
3) The user is redirected to a screen "myprojectname.firebaseapp.com" and then he is immediately redirect to a blank screen "about:blank"

There is a compatibility issue of the Safari web view?

I also opened a report to Firebase but we are still investigating.
We are looking for some help: the problem is really weird, specially because the behavior changes from time to time.

**Comment by MatthewPringle** [comment]
I am also seeing the about:blank issue

**Comment by MatthewPringle** [comment]
I have had the same issue, on iOS 12 at least of it never redirecting back to the app successfully.

Instead it eventually ends up at about:blank and the browser tab doesn't close.

It seems that the app goes to sleep when the sign in window is opened and isn't running when the callback is passed.

To solve this I installed https://github.com/katzer/cordova-plugin-background-mode and set it to keep the app awake in the background using cordova.plugins.backgroundMode.enable()

I plan to just enable and disable as needed but can anyone else check if it solves their problem.

**Comment by ancrezzi** [comment]
@MatthewPringle we tried to install the plugin but the app still doesn't work. When we navigate the application in background we don't see an instance of Safari with the "about:blank" page, but this is associated to our application. So it looks like your case is different from our.
Any other idea?

**Comment by MatthewPringle** [comment]
@ancrezzi I have managed to get it working on both Android and iOS

I am not using their recommended plugin Browsertabs as it was causing build issues on Android and seems to be abandoned.

Instead I am using cordova-plugin-safariviewcontroller

So what you need to do, before Firebase is loaded ( like the first thing after cordova loads ) is to replace the browsertab functions Firebase is hardcoded to look for with functions of the new plugin.

If you do it right, after Cordova Ready you should see a function attached to window.handleOpenURL by Firebase. This is the function that looks for the callback and activates the next internal step in Firebase Auth.

This is a work in progress but I have confirmed on live iOS and Android devices.

```
/* Browsertab */
        window.cordova.plugins.browsertab = {};
        
        /* Browsertab - Is Available */
        window.cordova.plugins.browsertab.isAvailable = function( success ) {
            window.SafariViewController.isAvailable( success );
        }
        
        /* Browsertab - Open Url */
        window.cordova.plugins.browsertab.openUrl = function( url ) {
            window.SafariViewController.show( { url: url } , function() {} , function() {} );
        }
        
        /* Browsertab - Close */
        window.cordova.plugins.browsertab.close = function() {
            window.SafariViewController.hide();
        }
```

```
"dependencies": {
    "cordova-android": "^7.1.4",
    "cordova-ios": "^4.5.5",
    "cordova-plugin-background-mode": "0.7.3",
    "cordova-plugin-buildinfo": "2.0.3",
    "cordova-plugin-customurlscheme": "4.4.0",
    "cordova-plugin-device": "2.0.3",
    "cordova-plugin-dialogs": "2.0.2",
    "cordova-plugin-inappbrowser": "^3.1.0",
    "cordova-plugin-network-information": "2.0.2",
    "cordova-plugin-safariviewcontroller": "^1.6.0",
    "cordova-plugin-screen-orientation": "3.0.2",
    "cordova-plugin-splashscreen": "5.0.3",
    "cordova-plugin-statusbar": "2.4.3",
    "cordova-plugin-whitelist": "1.3.4",
    "cordova-plugin-wkwebview-engine": "^1.2.1",
    "cordova-universal-links-plugin": "git+https://github.com/walteram/cordova-universal-links-plugin.git",
    "es6-promise-plugin": "4.2.2"
  }
```
